freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java

```diff
@@ -4,94 +4,225 @@
+import com.sequenceiq.freeipa.service.upgrade.ccm.UpgradeCcmOrchestratorService;
 
 @Service
 public class UpgradeCcmService {
 
+    private static final Logger LOGGER = LoggerFactory.getLogger(UpgradeCcmService.class);
+
     @Inject
     private StackUpdater stackUpdater;
 
-    public void checkPrerequsities(Long stackId) {
-        return;
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private ImageService imageService;
+
+    @Inject
+    private CcmUserDataService ccmUserDataService;
+
+    @Inject
+    private FreeIpaOrchestrationConfigService freeIpaOrchestrationConfigService;
+
+    @Inject
+    private BootstrapService bootstrapService;
+
+    @Inject
+    private UpgradeCcmOrchestratorService upgradeCcmOrchestratorService;
+
+    @Inject
+    private ClusterProxyService clusterProxyService;
+
+    @Inject
+    private FreeIpaStackHealthDetailsService healthService;
+
+    @Inject
+    private CcmResourceTerminationListener ccmResourceTerminationListener;
+
+    @Inject
+    private UpgradeService upgradeService;
+
+    public Stack checkPrerequsities(Long stackId) {
+        Stack stack = stackService.getStackById(stackId);
+        ImageEntity stackImage = imageService.getByStackId(stackId);
+        FreeIpaUpgradeOptions freeIpaUpgradeOptions = upgradeService.collectUpgradeOptions(stack.getAccountId(), stack.getEnvironmentCrn(),
+                Objects.requireNonNullElse(stackImage.getImageCatalogName(), stackImage.getImageCatalogUrl()));
+        if (!freeIpaUpgradeOptions.getImages().isEmpty()) {
+            throw new CloudbreakServiceException("FreeIPA is not on the latest available image. Please upgrade that first. CCM upgrade is not possible yet.");
+        }
+        return stack;
     }
 
     public void checkPrerequisitesState(Long stackId) {
         DetailedStackStatus detailedStatus = UPGRADE_CCM_IN_PROGRESS;
-        String statusReason = "Upgrade CCM in progress";
+        String statusReason = "Checking prerequisites";
         stackUpdater.updateStackStatus(stackId, detailedStatus, statusReason);
     }
 
-    public void finishedState(Long stackId) {
-        DetailedStackStatus detailedStatus = AVAILABLE;
-        String statusReason = "Upgrade CCM completed";
+    public void changeTunnelState(Long stackId) {
+        DetailedStackStatus detailedStatus = UPGRADE_CCM_IN_PROGRESS;
+        String statusReason = "Changing tunnel type in database";
         stackUpdater.updateStackStatus(stackId, detailedStatus, statusReason);
     }
 
-    public void failedState(Long stackId) {
-        DetailedStackStatus detailedStatus = UPGRADE_CCM_FAILED;
-        String statusReason = "Upgrade CCM failed";
-        stackUpdater.updateStackStatus(stackId, detailedStatus, statusReason);
+    public void changeTunnel(Long stackId) {
+        Stack stack = stackService.getStackById(stackId);
+        stack.setTunnel(Tunnel.latestUpgradeTarget());
+        stackService.save(stack);
```

Comment:
instead of fetching and saving the whole entity you could use a special `@Modifying` annotated query to update this field

Reply:
Ok. I did the same in the CcmUserDataService, for saving CCMv2 agent crn